### PR TITLE
ui/qa: hook up question submit component + refactor

### PR DESCRIPTION
**Pull request by b1lly**
- creates a core model for the qa submission component, and decouples it from results storage dependency

- component/template handles a few various states for the component (errors, submit successful, and results)

- attempt a new pattern for default config component. with this pattern, if components need to reference config, its through `this._config.[prop]`. templates continue to reference through `_config.[prop]`. 

default values are provided via a separate object and not defined explicitly on the constructor.  this helps gets rid of a bunch of boilerplate code between the constructor assignment and setstate for passing configuration to templates. It also can unify how we handle configuration (between definition and validation).  

- because of the way the default config can work, added a validateConfig method that each component sub class can have for throwing errors relative to the component. in the future, this could be an extension of the base class if we like the pattern

### CODE REVIEW

File: src/ui/components/questions/questionsubmissioncomponent.js [hunk 1]
```diff
@@ -60,49 +91,103 @@ export default class QuestionSubmissionComponent extends Component {
...
-      emailLabel: this._emailLabel,
-      nameLabel: this._nameLabel,
-      questionLabel: this._questionLabel,
-      privacyPolicyLabel: this._privacyPolicyLabel,
-      buttonLabel: this._buttonLabel,
-      question: new SearchParams(window.location.search.substring(1)).get('query')
-    });
+  /**
+   * Validates the required fields (or rules) for the formData
+   * @param {HTMLElement} formData
```
Comment (b1lly) [inline src/ui/components/questions/questionsubmissioncomponent.js]: This actually is an {object} and not and HTMLElement. I'll fix this tomorrow.

File: src/core/core.js [hunk 2]
```diff
@@ -90,33 +91,78 @@ export default class Core {
...
+      })
       .then(response => SearchDataTransformer.transformVertical(response))
       .then(results => query.append
         ? this.storage.getState(StorageKeys.VERTICAL_RESULTS).append(results)
         : results)
       .then(data => {
-        this.storage.set(StorageKeys.QUERY_ID, data[StorageKeys.QUERY_ID]);
-        this.storage.set(StorageKeys.NAVIGATION, data[StorageKeys.NAVIGATION]);
-        this.storage.set(StorageKeys.VERTICAL_RESULTS, data[StorageKeys.VERTICAL_RESULTS]);
+        this.storage.set(
```
Comment (Delerme) [inline src/core/core.js]: I find this a lot harder to read and follow, can we avoid multi-lining expressions if they are under 120 characters?

File: src/core/core.js [hunk 3]
```diff
@@ -90,33 +91,78 @@ export default class Core {
...
         if (data[StorageKeys.DYNAMIC_FILTERS]) {
-          this.storage.set(StorageKeys.DYNAMIC_FILTERS, data[StorageKeys.DYNAMIC_FILTERS]);
+          this.storage.set(StorageKeys.DYNAMIC_FILTERS,
+            data[StorageKeys.DYNAMIC_FILTERS]);
         }
+
+        this.storage.set(
+          StorageKeys.QUESTION_SUBMISSION,
+          new QuestionSubmission({
+            questionText: query.input
```
Comment (benmcginnis) [inline src/core/core.js]: this will default the question text to the value of the user's query, correct?

File: src/core/storage/storagekeys.js [hunk 4]
```diff
@@ -16,5 +16,6 @@ export default {
   QUERY: 'query',
   QUERY_ID: 'query-id',
   FACET_FILTER: 'facet-filter',
-  DYNAMIC_FILTERS: 'dynamic-filters'
+  DYNAMIC_FILTERS: 'dynamic-filters',
+  QUESTION_SUBMISSION: 'question-submission'
```
Comment (benmcginnis) [inline src/core/storage/storagekeys.js:20]: you should be able to use a trailing comma here, to avoid two line diffs when you add a row in the future.

File: src/ui/components/questions/questionsubmissioncomponent.js [hunk 5]
```diff
@@ -4,54 +4,85 @@ import Component from '../component';
...
-     * Optionally provided, otherwise defaults to `Name`
-     * @type {string}
-     */
-    this._nameLabel = opts.nameLabel || 'Name:';
+  /**
+   * The label to use for the name input
+   * Optionally provided, otherwise defaults to `Name`
+   * @type {string}
+   */
+  'nameLabel': '*Name:',
```
Comment (benmcginnis) [inline src/ui/components/questions/questionsubmissioncomponent.js]: I think you want to separate the * from the actual label.

File: src/ui/components/questions/questionsubmissioncomponent.js (hunk 1 above)
Comment (benmcginnis) [reply src/ui/components/questions/questionsubmissioncomponent.js]: Does JSDoc respect the actual dom api models? If so, you might want to use {FormData}: https://developer.mozilla.org/en-US/docs/Web/API/FormData

File: src/ui/templates/questions/questionsubmission.hbs [hunk 6]
```diff
@@ -0,0 +1,46 @@
+{{#if questionSubmitted}}
+  <div class="yxt-QuestionSubmission">Thanks for your submission!</div>
+{{else}}
+  <form class="yxt-QuestionSubmission">
+    {{#if errors}}
+      <div class="yxt-errors">
+        You're missing some required fields!
```
Comment (benmcginnis) [inline src/ui/templates/questions/questionsubmission.hbs]: I think you'd also want to add  the errors you made, right? Perhaps as a bulleted list?

**Review by benmcginnis — APPROVED** [review]

File: src/core/core.js (hunk 2 above)
Comment (b1lly) [reply src/core/core.js]: Yes! Apologies. I made the adjustment already just didn't re-upload.

File: src/core/core.js (hunk 3 above)
Comment (b1lly) [reply src/core/core.js]: Yeah, that was the original behavior Max wanted. Happy to change it though.

File: src/ui/templates/questions/questionsubmission.hbs (hunk 6 above)
Comment (b1lly) [reply src/ui/templates/questions/questionsubmission.hbs]: Yeah, I left this simple because I wasn't exactly sure how we wanted to handle the errors. I kind've just plopped it in here to make sure the mechanism existed and worked.

For instance, there's a few ways:

1. Bulleted list
2. Each input appends an `yxt-error` class to highlight them red
3. Each input shows the error message above/below

Do we know of a preference?

File: src/core/core.js (hunk 2 above)
Comment (b1lly) [reply src/core/core.js]: Do you know if this is something we can bake into the formatter? @Delerme

File: src/core/storage/storagekeys.js (hunk 4 above)
Comment (b1lly) [reply src/core/storage/storagekeys.js:20]: Yeah, I grew up with the trailing comma practice too, but the formatter we're using trims them unfortunately.

File: src/ui/templates/questions/questionsubmission.hbs (hunk 6 above)
Comment (benmcginnis) [reply src/ui/templates/questions/questionsubmission.hbs]: I can take a crack at making it nice & accessible, but this is a good overview of the best practice: https://www.w3.org/WAI/tutorials/forms/notifications/

TL;DR:

```
<div class="yxt-errors" role="alert">
  <h3>You are missing {{errors.length}} required fields</h3>
  <ul>
      {{#each errors}}
          <li>{{error}}</li>
      {{/each}}
   </ul>
</div>
```

File: src/ui/components/questions/questionsubmissioncomponent.js (hunk 5 above)
Comment (b1lly) [reply src/ui/components/questions/questionsubmissioncomponent.js]: Sure. Updated markup.

File: src/core/core.js (hunk 3 above)
Comment (benmcginnis) [reply src/core/core.js]: just wanted to confirm as Rose asked me about making it a default

File: src/ui/templates/questions/questionsubmission.hbs (hunk 6 above)
Comment (b1lly) [reply src/ui/templates/questions/questionsubmission.hbs]: added.

**Comment by benmcginnis** [comment]
@b1lly, @creotutar pointed out to me that the analytics spec wants us to capture events whenever anyone focuses on the inputs in this form.  Do you want to add that as part of this PR or do another one?

File: .size-limit.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 module.exports = [
   {
     path: 'dist/answers.min.js',
-    limit: '17 KB'
+    limit: '25 KB'
```
Comment (benmcginnis) [inline .size-limit.js:4]: 😢

Comment (b1lly) [reply .size-limit.js:4]: > @b1lly, @creotutar pointed out to me that the analytics spec wants us to capture events whenever anyone focuses on the inputs in this form. Do you want to add that as part of this PR or do another one?

I'm actually happy to implement now. Are we doing it on FOCUS or on click? I think there's a difference. What type of data are we actually capturing?

'User on input'?

Or is it 'User on email input'?

Can someone DM me a link on slack with the spec?

**Comment by benmcginnis** [comment]
Focus should fire for both click and tabbing into the element, so I think
we want focus

File: src/core/core.js (hunk 2 above)
Comment (Delerme) [reply src/core/core.js]: don't think so, there's no real hard and fast rule for multiline vs single, it's basically just whatever subjectively looks good, so you can't really represent that in a linter

File: src/core/core.js [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 import SearchApi from './search/searchapi';
 import AutoCompleteApi from './search/autocompleteapi';
+import QuestionAnswerApi from './search/questionanswerapi';
```
Comment (b1lly) [inline src/core/core.js:5]: I plan on moving all of these into an `api` or `network` package.

File: src/core/core.js (hunk 2 above)
Comment (b1lly) [reply src/core/core.js]: Ah poops. Doesn't the Java style thing do something like this? It's not always perfect but it tries to auto format and allows you to tweak it if its not ideal.

**Comment by Delerme** [comment]
is this ready for review? take out of WIP status if so

**Comment by b1lly** [comment]
> is this ready for review? take out of WIP status if so

Yup, should be in a good enough spot to review. I do want to add tests so wasn't sure if it should come out of WIP yet. There's a minor tweak also to how language should be provided for the component (waiting for product).

**Review by benmcginnis — APPROVED** [review]

**Comment by b1lly** [comment]
This is done and ready to be reviewed. Sorry. I'll add tests separately.

**Review by benmcginnis — APPROVED** [review]
lgtm

File: src/core/core.js [hunk 9]
```diff
@@ -76,6 +78,15 @@ export default class Core {
       answersKey: this._answersKey,
       locale: this._locale
     });
+
+    /**
+     * An abstraction
```
Comment (Delerme) [inline src/core/core.js]: ?

File: src/core/core.js [hunk 10]
```diff
@@ -90,33 +101,64 @@ export default class Core {
    * @param {boolean} query.append If true, adds the results of this query to the end of the current results, defaults false
    */
   verticalSearch (verticalKey, query) {
-    this.storage.set(StorageKeys.VERTICAL_RESULTS, VerticalResults.searchLoading());
+    this.storage.set(
```
Comment (Delerme) [inline src/core/core.js]: single line

File: src/core/core.js [hunk 11]
```diff
@@ -90,33 +101,64 @@ export default class Core {
...
+      })
       .then(response => SearchDataTransformer.transformVertical(response))
       .then(results => query.append
         ? this.storage.getState(StorageKeys.VERTICAL_RESULTS).append(results)
         : results)
       .then(data => {
         this.storage.set(StorageKeys.QUERY_ID, data[StorageKeys.QUERY_ID]);
         this.storage.set(StorageKeys.NAVIGATION, data[StorageKeys.NAVIGATION]);
-        this.storage.set(StorageKeys.VERTICAL_RESULTS, data[StorageKeys.VERTICAL_RESULTS]);
+        this.storage.set(
```
Comment (Delerme) [inline src/core/core.js]: single line

File: src/core/core.js [hunk 12]
```diff
@@ -90,33 +101,64 @@ export default class Core {
...
+          StorageKeys.VERTICAL_RESULTS,
+          data[StorageKeys.VERTICAL_RESULTS]);
+
         if (data[StorageKeys.DYNAMIC_FILTERS]) {
-          this.storage.set(StorageKeys.DYNAMIC_FILTERS, data[StorageKeys.DYNAMIC_FILTERS]);
+          this.storage.set(
```
Comment (Delerme) [inline src/core/core.js]: single line

File: src/core/core.js [hunk 13]
```diff
@@ -90,33 +101,64 @@ export default class Core {
...
+          new QuestionSubmission({
+            questionText: query.input
+          })
+        );
       });
   }
 
   search (queryString, urls) {
-    this.storage.set(StorageKeys.UNIVERSAL_RESULTS, UniversalResults.searchLoading());
+    this.storage.set(
```
Comment (Delerme) [inline src/core/core.js]: single line

File: src/core/core.js [hunk 14]
```diff
@@ -167,6 +209,27 @@ export default class Core {
       });
   }
 
+  /**
+   * Submits a question to the server and updates the underlying questin model
```
Comment (Delerme) [inline src/core/core.js]: typo

File: src/core/models/questionsubmission.js [hunk 15]
```diff
@@ -0,0 +1,53 @@
+/** @module QuestionSubmission */
+
+export default class QuestionSubmission {
```
Comment (Delerme) [inline src/core/models/questionsubmission.js:7]: jsdoc

File: src/core/search/questionanswerapi.js [hunk 16]
```diff
@@ -0,0 +1,47 @@
+/** @module QuestionAnswerApi */
+
+import ApiRequest from '../http/apirequest';
+import { AnswersBasicError, AnswersEndpointError } from '../errors/errors';
+
+/**
+ * QuestionAnswerApi exposes an interface for network related matters
+ * for all the autocomplete endpoints.
```
Comment (Delerme) [inline src/core/search/questionanswerapi.js]: ?

File: src/core/search/questionanswerapi.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+     * @type {string}
+     * @private
+     */
+    if (!opts.apiKey) {
+      throw new AnswersBasicError('Api Key is required', 'QuestionAnswerApi');
+    }
+    this._apiKey = opts.apiKey;
+  }
+
+  submitQuestion (question) {
```
Comment (Delerme) [inline src/core/search/questionanswerapi.js:35]: jsdoc

File: src/ui/components/questions/questionsubmissioncomponent.js [hunk 18]
```diff
@@ -3,55 +3,86 @@
...
+/**
+ * Configurable options for the component
+ * @type {Object}
+ */
+const DEFAULT_CONFIG = {
+  /**
+   * The entity identifier that the question is associated with.
+   * This is typically an organization object
+   * @type {number}
+   * @required
```
Comment (Delerme) [inline src/ui/components/questions/questionsubmissioncomponent.js]: I don't think this exists

File: src/ui/components/questions/questionsubmissioncomponent.js [hunk 19]
```diff
@@ -60,49 +91,125 @@ export default class QuestionSubmissionComponent extends Component {
...
+      return;
+    }
+
+    this.bindAnalytics(formEl);
+    this.bindFormSubmit(formEl);
   }
 
-  initSubmit (formSelector) {
-    this._formEl = formSelector;
+  bindAnalytics (formEl) {
```
Comment (Delerme) [inline src/ui/components/questions/questionsubmissioncomponent.js:138]: jsdoc

File: src/ui/components/questions/questionsubmissioncomponent.js [hunk 20]
```diff
@@ -60,49 +91,125 @@ export default class QuestionSubmissionComponent extends Component {
...
-    return false;
+  /**
+   * validationConfig contains a bunch of rules
+   * that are used to validate aginst configuration provided by the user
+   */
+  validateConfig () {
+    if (this._config.entityId === null ||
+        typeof this._config.entityId !== 'number') {
+      throw new AnswersComponentError(
+        '`entityId` is a required configuration option for Question Submission',
```
Comment (benmcginnis) [inline src/ui/components/questions/questionsubmissioncomponent.js:113]: Wait, this is strange, how do we know the entity Id for the question submission?

File: src/ui/components/questions/questionsubmissioncomponent.js [hunk 21]
```diff
@@ -60,49 +91,125 @@ export default class QuestionSubmissionComponent extends Component {
...
+        });
     });
   }
 
-  static get type () {
-    return 'QASubmission';
+  /**
+   * Takes the form, and builds a object that represents the input names
+   * and text fields.
+   * @param {HTMLElement} formData
```
Comment (Delerme) [inline src/ui/components/questions/questionsubmissioncomponent.js]: formEl

File: src/ui/components/questions/questionsubmissioncomponent.js [hunk 22]
```diff
@@ -60,49 +91,125 @@ export default class QuestionSubmissionComponent extends Component {
...
+
+    if (!formData.privacyPolicy || formData.privacyPolicy !== true) {
+      errors['privacyPolicy'] = 'Approving our privacy policy terms is required!';
+    }
+
+    if (!formData.questionText || typeof formData.questionText !== 'string') {
+      errors['questionText'] = 'Question text is required!';
+    }
+
+    return Object.keys(errors).length > 0 ? errors : false;
```
Comment (Delerme) [inline src/ui/components/questions/questionsubmissioncomponent.js]: does it make more sense to just return an empty object?

**Review by Delerme — CHANGES_REQUESTED** [review]

File: src/ui/components/questions/questionsubmissioncomponent.js (hunk 20 above)
Comment (b1lly) [reply src/ui/components/questions/questionsubmissioncomponent.js:113]: Entity ID in this case is the `organization` object which represents the brand. It's a single object that the account will have and will always be the one that you provide for the business/brand.

The point of this, is that it makes the downstream systems that depend on entities for things to work more easily as their primary relationship object.

File: src/ui/components/questions/questionsubmissioncomponent.js (hunk 22 above)
Comment (b1lly) [reply src/ui/components/questions/questionsubmissioncomponent.js]: Probably not. I don't really feel strongly though. 

Otherwise you'd have to inspect it for errors.
I was doing this based off how `golang` handles errors for instance; the pattern there is to return an error, or `nil`.

let errors = checkErrors(stuff);
if (errors) {
  // do something
}

**Comment by benmcginnis** [comment]
Great, thanks for the clarification.

File: src/ui/components/questions/questionsubmissioncomponent.js (hunk 18 above)
Comment (b1lly) [reply src/ui/components/questions/questionsubmissioncomponent.js]: I'm an innovator.

File: src/ui/components/questions/questionsubmissioncomponent.js (hunk 22 above)
Comment (Delerme) [reply src/ui/components/questions/questionsubmissioncomponent.js]: maybe null then? Just seems strange to have a return type that's either an object or a boolean. I get the pattern and although it's prevalent in Go, I don't think anyone in JS expects something like that

File: src/ui/components/questions/questionsubmissioncomponent.js (hunk 18 above)
Comment (Delerme) [reply src/ui/components/questions/questionsubmissioncomponent.js]: I know you are,  but let's save the innovation for another day

**Review by Delerme — APPROVED** [review]
